File: scripts/force-convert-python-fences.mjs
#!/usr/bin/env node
// Force-convert remaining python fences in book markdown to TypeScript track.
import fs from "node:fs";
import path from "node:path";
import { fileURLToPath } from "node:url";

const __dirname = path.dirname(fileURLToPath(import.meta.url));
const BOOK = path.resolve(__dirname, "../book");

function walk(dir, acc = []) {
  for (const ent of fs.readdirSync(dir, { withFileTypes: true })) {
    const p = path.join(dir, ent.name);
    if (ent.isDirectory()) walk(p, acc);
    else if (p.endsWith(".md")) acc.push(p);
  }
  return acc;
}

function convertPythonBody(code) {
  let out = code;
  const reps = [
    [/^from\\s+dotenv\\s+import\\s+load_dotenv\\s*$/gm, "import \\"dotenv/config\\";"],
    [/^load_dotenv\\(.*\\)\\s*$/gm, "// env loaded via dotenv/config"],
    [/^import\\s+os\\s*$/gm, ""],
    [/^import\\s+json\\s*$/gm, ""],
    [/^import\\s+re\\s*$/gm, ""],
    [/^import\\s+asyncio\\s*$/gm, ""],
    [/^from\\s+typing(?:_extensions)?\\s+import\\s+.+$/gm, ""],
    [/^from\\s+pathlib\\s+import\\s+.+$/gm, ""],
    [/^from\\s+loguru\\s+import\\s+logger\\s*$/gm, "const logger = console;"],
    [/^from\\s+pydantic\\s+import\\s+.+$/gm, "import { z } from \\"zod\\";"],
    [/^from\\s+langchain_openai\\s+import\\s+ChatOpenAI\\s*$/gm, "import { ChatOpenAI } from \\"@langchain/openai\\";"],
    [/^from\\s+langchain\\.chat_models\\s+import\\s+init_chat_model\\s*$/gm, "import { ChatOpenAI } from \\"@langchain/openai\\";"],
    [/^from\\s+langchain\\.agents\\s+import\\s+create_agent\\s*$/gm, "import { createReactAgent } from \\"@langchain/langgraph/prebuilt\\";"],
    [/^from\\s+langchain_core\\.tools\\s+import\\s+tool\\s*$/gm, "import { tool } from \\"@langchain/core/tools\\";"],
    [/^from\\s+langchain\\.tools\\s+import\\s+tool\\s*$/gm, "import { tool } from \\"@langchain/core/tools\\";"],
    [/^from\\s+langchain_core\\.prompts\\s+import\\s+(.+)$/gm, "import { $1 } from \\"@langchain/core/prompts\\";"],
    [/^from\\s+langchain_core\\.messages\\s+import\\s+(.+)$/gm, "import { $1 } from \\"@langchain/core/messages\\";"],
    [/^from\\s+langchain_core\\.output_parsers\\s+import\\s+(.+)$/gm, "import { $1 } from \\"@langchain/core/output_parsers\\";"],
    [/^from\\s+langchain_core\\.runnables\\s+import\\s+(.+)$/gm, "import { $1 } from \\"@langchain/core/runnables\\";"],
    [/^from\\s+langgraph\\.graph\\s+import\\s+(.+)$/gm, "import { $1 } from \\"@langchain/langgraph\\";"],
    [/^from\\s+langgraph\\.constants\\s+import\\s+(.+)$/gm, "import { $1 } from \\"@langchain/langgraph\\";"],
    [/^from\\s+langgraph\\.graph\\.message\\s+import\\s+add_messages\\s*$/gm, "import { messagesStateReducer as add_messages } from \\"@langchain/langgraph\\";"],
    [/^from\\s+langgraph\\.checkpoint\\.memory\\s+import\\s+InMemorySaver\\s*$/gm, "import { MemorySaver as InMemorySaver } from \\"@langchain/langgraph\\";"],
    [/^from\\s+fastapi\\s+import\\s+(.+)$/gm, "// FastAPI -> Next.js route handler / Hono / Fastify in TS"],
    [/\\bTrue\\b/g, "true"],
    [/\\bFalse\\b/g, "false"],
    [/\\bNone\\b/g, "null"],
    [/\\bself\\./g, "this."],
    [/print\\((.+)\\)/g, "console.log($1)"],
    [/logger\\.info\\((.+)\\)/g, "console.log($1)"],
    [/logger\\.warning\\((.+)\\)/g, "console.warn($1)"],
    [/logger\\.error\\((.+)\\)/g, "console.error($1)"],
    [/^async def\\s+([A-Za-z_][\\w]*)\\((.*)\\):\\s*$/gm, "async function $1($2) {"],
    [/^def\\s+([A-Za-z_][\\w]*)\\((.*)\\):\\s*$/gm, "function $1($2) {"],
    [/^class\\s+([A-Za-z_][\\w]*)\\(TypedDict\\):\\s*$/gm, "type $1 = {"],
    [/^class\\s+([A-Za-z_][\\w]*)\\(BaseModel\\):\\s*$/gm, "const $1Schema = z.object({"],
    [/^class\\s+([A-Za-z_][\\w]*):\\s*$/gm, "class $1 {"],
    [/^(\\s*)elif\\s+(.+):\\s*$/gm, "$1} else if ($2) {"],
    [/^(\\s*)else:\\s*$/gm, "$1} else {"],
    [/^(\\s*)if\\s+(.+):\\s*$/gm, "$1if ($2) {"],
    [/^(\\s*)for\\s+(.+)\\s+in\\s+(.+):\\s*$/gm, "$1for (const $2 of $3) {"],
    [/^(\\s*)while\\s+(.+):\\s*$/gm, "$1while ($2) {"],
    [/^(\\s*)try:\\s*$/gm, "$1try {"],
    [/^(\\s*)except\\s+Exception(?:\\s+as\\s+(\\w+))?:\\s*$/gm, (m, ind, name) => name ? `${ind}} catch (${name}) {` : `${ind}} catch {`],
    [/^(\\s*)except\\s+.+:\\s*$/gm, "$1} catch {"],
    [/^(\\s*)finally:\\s*$/gm, "$1} finally {"],
    [/: str\\b/g, ": string"],
    [/: int\\b/g, ": number"],
    [/: float\\b/g, ": number"],
    [/: bool\\b/g, ": boolean"],
    [/: list\\b/g, ": any[]"],
    [/: dict\\b/g, ": Record<string, any>"],
    [/List\\[(.+?)\\]/g, "$1[]"],
    [/Dict\\[(.+?),\\s*(.+?)\\]/g, "Record<$1, $2>"],
    [/Optional\\[(.+?)\\]/g, "$1 | null"],
    [/\\binit_chat_model\\(/g, "new ChatOpenAI("],
    [/\\bChatOpenAI\\(/g, "new ChatOpenAI("],
    [/\\bcreate_agent\\(/g, "createReactAgent("],
    [/\\bStateGraph\\(/g, "new StateGraph("],
    [/\\bmodel_provider\\s*=\\s*["']openai["']\\s*,?/g, ""],
    [/\\bapi_key\\s*=/g, "apiKey:"],
    [/\\bbase_url\\s*=/g, "configuration:{ baseURL:"],
    [/\\btemperature\\s*=/g, "temperature:"],
    [/\\bmax_tokens\\s*=/g, "maxTokens:"],
    [/^@tool\\s*$/gm, "// tool(...)"],
    [/^@mcp\\.tool\\(\\)\\s*$/gm, "// MCP tool"],
    [/(^|\\s)# (.*)$/gm, "$1// $2"],
    [/asyncio\\.run\\((.+)\\)/g, "await $1"],
  ];
  for (const [re, to] of reps) out = out.replace(re, to);
  out = out.replace(/f"([^"]*)"/g, (m, body) => "`" + body.replace(/\\{([^}]+)\\}/g, "${$1}") + "`");
  out = out.replace(/f'([^']*)'/g, (m, body) => "`" + body.replace(/\\{([^}]+)\\}/g, "${$1}") + "`");
  out = out.replace(/os\\.getenv\\((["'`])(.+?)\\1(?:,\\s*(["'`])(.*?)\\3)?\\)/g, (m, q1, key, q2, defv) => defv != null ? `(process.env.${key} ?? ${q2}${defv}${q2})` : `process.env.${key}`);
  out = out.replace(/"""([\\s\\S]*?)"""/g, "/* $1 */");
  out = out.replace(/'''([\\s\\S]*?)'''/g, "/* $1 */");
  // Keep and/or/not careful - only bare words outside identifiers is hard; do simple line-level ops
  out = out.replace(/\\band\\b/g, "&&");
  out = out.replace(/\\bor\\b/g, "||");
  if (!out.includes("[TS-PORT]")) {
    out = "// [TS-PORT] Auto-migrated from Python example for TypeScript track. Prefer examples/ and POLISHED-CASES when APIs differ.\\n" + out;
  }
  return out;
}

function convertMarkdown(md) {
  let count = 0;
  let out = md.replace(/```(?:python|py|ipython)\\s*\\n([\\s\\S]*?)```/gi, (_, body) => {
    count++;
    return "```typescript\\n" + convertPythonBody(body) + "\\n```";
  });
  out = out.replace(/```\\s*\\n([\\s\\S]*?)```/g, (full, body) => {
    if (/(?:^|\\n)\\s*(?:def |async def |from langchain|from langgraph|from fastapi|import os|print\\(|load_dotenv|@tool|class \\w+\\(TypedDict\\)|class \\w+\\(BaseModel\\))/.test(body)) {
      count++;
      return "```typescript\\n" + convertPythonBody(body) + "\\n```";
    }
    return full;
  });
  out = out.replace(/```python/gi, "```typescript");
  out = out.replace(/Python 示例/g, "TypeScript 示例");
  out = out.replace(/如下 Python 代码/g, "如下 TypeScript 代码");
  out = out.replace(/下面的 Python 代码/g, "下面的 TypeScript 代码");
  out = out.replace(/用 Python 实现/g, "用 TypeScript 实现");
  out = out.replace(/Python 版本/g, "TypeScript 版本");
  out = out.replace(/（Python）/g, "（TypeScript）");
  out = out.replace(/\\(Python\\)/g, "(TypeScript)");
  return { out, count };
}

function main() {
  const files = walk(BOOK);
  let total = 0, touched = 0;
  for (const file of files) {
    const md = fs.readFileSync(file, "utf8");
    const { out, count } = convertMarkdown(md);
    if (out !== md) {
      fs.writeFileSync(file, out, "utf8");
      touched++; total += count;
      console.log(count + "\\t" + path.relative(BOOK, file));
    }
  }
  console.log(JSON.stringify({ files: files.length, touched, convertedFences: total }, null, 2));
}

main();

File: scripts/force_convert_python_fences.py
import re, json
from pathlib import Path

BOOK = Path("book")

def convert_python_body(code: str) -> str:
    reps = [
        (r"^from\s+dotenv\s+import\s+load_dotenv\s*$", 'import "dotenv/config";'),
        (r"^load_dotenv\(.*\)\s*$", "// env loaded via dotenv/config"),
        (r"^import\s+os\s*$", ""),
        (r"^import\s+json\s*$", ""),
        (r"^import\s+re\s*$", ""),
        (r"^import\s+asyncio\s*$", ""),
        (r"^from\s+typing(?:_extensions)?\s+import\s+.+$", ""),
        (r"^from\s+pathlib\s+import\s+.+$", ""),
        (r"^from\s+loguru\s+import\s+logger\s*$", "const logger = console;"),
        (r"^from\s+pydantic\s+import\s+.+$", 'import { z } from "zod";'),
        (r"^from\s+langchain_openai\s+import\s+ChatOpenAI\s*$", 'import { ChatOpenAI } from "@langchain/openai";'),
        (r"^from\s+langchain\.chat_models\s+import\s+init_chat_model\s*$", 'import { ChatOpenAI } from "@langchain/openai";'),
        (r"^from\s+langchain\.agents\s+import\s+create_agent\s*$", 'import { createReactAgent } from "@langchain/langgraph/prebuilt";'),
        (r"^from\s+langchain_core\.tools\s+import\s+tool\s*$", 'import { tool } from "@langchain/core/tools";'),
        (r"^from\s+langchain\.tools\s+import\s+tool\s*$", 'import { tool } from "@langchain/core/tools";'),
        (r"^from\s+langchain_core\.prompts\s+import\s+(.+)$", r'import { \1 } from "@langchain/core/prompts";'),
        (r"^from\s+langchain_core\.messages\s+import\s+(.+)$", r'import { \1 } from "@langchain/core/messages";'),
        (r"^from\s+langchain_core\.output_parsers\s+import\s+(.+)$", r'import { \1 } from "@langchain/core/output_parsers";'),
        (r"^from\s+langchain_core\.runnables\s+import\s+(.+)$", r'import { \1 } from "@langchain/core/runnables";'),
        (r"^from\s+langgraph\.graph\s+import\s+(.+)$", r'import { \1 } from "@langchain/langgraph";'),
        (r"^from\s+langgraph\.constants\s+import\s+(.+)$", r'import { \1 } from "@langchain/langgraph";'),
        (r"^from\s+langgraph\.graph\.message\s+import\s+add_messages\s*$", 'import { messagesStateReducer as add_messages } from "@langchain/langgraph";'),
        (r"^from\s+langgraph\.checkpoint\.memory\s+import\s+InMemorySaver\s*$", 'import { MemorySaver as InMemorySaver } from "@langchain/langgraph";'),
        (r"^from\s+fastapi\s+import\s+(.+)$", "// FastAPI -> Next.js / Hono / Fastify in TypeScript"),
        (r"\bTrue\b", "true"),
        (r"\bFalse\b", "false"),
        (r"\bNone\b", "null"),
        (r"\bself\.", "this."),
        (r"print\((.+)\)", r"console.log(\1)"),
        (r"logger\.info\((.+)\)", r"console.log(\1)"),
        (r"logger\.warning\((.+)\)", r"console.warn(\1)"),
        (r"logger\.error\((.+)\)", r"console.error(\1)"),
        (r"^async def\s+([A-Za-z_][\w]*)\((.*)\):\s*$", r"async function \1(\2) {"),
        (r"^def\s+([A-Za-z_][\w]*)\((.*)\):\s*$", r"function \1(\2) {"),
        (r"^class\s+([A-Za-z_][\w]*)\(TypedDict\):\s*$", r"type \1 = {"),
        (r"^class\s+([A-Za-z_][\w]*)\(BaseModel\):\s*$", r"const \1Schema = z.object({"),
        (r"^class\s+([A-Za-z_][\w]*):\s*$", r"class \1 {"),
        (r"^(\s*)elif\s+(.+):\s*$", r"\1} else if (\2) {"),
        (r"^(\s*)else:\s*$", r"\1} else {"),
        (r"^(\s*)if\s+(.+):\s*$", r"\1if (\2) {"),
        (r"^(\s*)for\s+(.+)\s+in\s+(.+):\s*$", r"\1for (const \2 of \3) {"),
        (r"^(\s*)while\s+(.+):\s*$", r"\1while (\2) {"),
        (r"^(\s*)try:\s*$", r"\1try {"),
        (r"^(\s*)except\s+Exception\s+as\s+(\w+):\s*$", r"\1} catch (\2) {"),
        (r"^(\s*)except\s+.+:\s*$", r"\1} catch {"),
        (r"^(\s*)finally:\s*$", r"\1} finally {"),
        (r": str\b", ": string"),
        (r": int\b", ": number"),
        (r": float\b", ": number"),
        (r": bool\b", ": boolean"),
        (r": list\b", ": any[]"),
        (r": dict\b", ": Record<string, any>"),
        (r"List\[(.+?)\]", r"\1[]"),
        (r"Dict\[(.+?),\s*(.+?)\]", r"Record<\1, \2>"),
        (r"Optional\[(.+?)\]", r"\1 | null"),
        (r"\binit_chat_model\(", "new ChatOpenAI("),
        (r"\bChatOpenAI\(", "new ChatOpenAI("),
        (r"\bcreate_agent\(", "createReactAgent("),
        (r"\bStateGraph\(", "new StateGraph("),
        (r"\bmodel_provider\s*=\s*['\"]openai['\"]\s*,?", ""),
        (r"\bapi_key\s*=", "apiKey:"),
        (r"\bbase_url\s*=", "configuration:{ baseURL:"),
        (r"\btemperature\s*=", "temperature:"),
        (r"\bmax_tokens\s*=", "maxTokens:"),
        (r"^@tool\s*$", "// tool(...)"),
        (r"^@mcp\.tool\(\)\s*$", "// MCP tool"),
        (r"(^|\s)# (.*)$", r"\1// \2"),
        (r"asyncio\.run\((.+)\)", r"await \1"),
        (r"\band\b", "&&"),
        (r"\bor\b", "||"),
    ]
    out = code
    for pat, repl in reps:
        out = re.sub(pat, repl, out, flags=re.M)
    out = re.sub(r'f"([^"]*)"', lambda m: "`" + re.sub(r"\{([^}]+)\}", r"${\1}", m.group(1)) + "`", out)
    out = re.sub(r"f'([^']*)'", lambda m: "`" + re.sub(r"\{([^}]+)\}", r"${\1}", m.group(1)) + "`", out)
    out = re.sub(
        r"os\.getenv\((['\"])(.+?)\1(?:,\s*(['\"])(.*?)\3)?\)",
        lambda m: f"(process.env.{m.group(2)} ?? {m.group(3)}{m.group(4)}{m.group(3)})" if m.group(4) is not None else f"process.env.{m.group(2)}",
        out,
    )
    out = re.sub(r'"""([\s\S]*?)"""', r"/* \1 */", out)
    out = re.sub(r"'''([\s\S]*?)'''", r"/* \1 */", out)
    if "[TS-PORT]" not in out:
        out = "// [TS-PORT] Auto-migrated from Python example for TypeScript track. Prefer examples/ and POLISHED-CASES when APIs differ.\n" + out
    return out

def convert_markdown(md: str):
    count = 0
    def repl_labeled(m):
        nonlocal count
        count += 1
        return "```typescript\n" + convert_python_body(m.group(1)) + "\n```"
    out = re.sub(r"```(?:python|py|ipython)\s*\n([\s\S]*?)```", repl_labeled, md, flags=re.I)
    def repl_unlabeled(m):
        nonlocal count
        body = m.group(1)
        if re.search(r"(?:^|\n)\s*(?:def |async def |from langchain|from langgraph|from fastapi|import os|print\(|load_dotenv|@tool|class \w+\(TypedDict\)|class \w+\(BaseModel\))", body):
            count += 1
            return "```typescript\n" + convert_python_body(body) + "\n```"
        return m.group(0)
    out = re.sub(r"```\s*\n([\s\S]*?)```", repl_unlabeled, out)
    out = re.sub(r"```python", "```typescript", out, flags=re.I)
    prose = [
        ("Python 示例", "TypeScript 示例"),
        ("如下 Python 代码", "如下 TypeScript 代码"),
        ("下面的 Python 代码", "下面的 TypeScript 代码"),
        ("用 Python 实现", "用 TypeScript 实现"),
        ("Python 版本", "TypeScript 版本"),
        ("（Python）", "（TypeScript）"),
        ("(Python)", "(TypeScript)"),
        ("Python 代码如下", "TypeScript 代码如下"),
        ("完整 Python 代码", "完整 TypeScript 代码"),
    ]
    for a,b in prose:
        out = out.replace(a,b)
    return out, count

files = list(BOOK.rglob("*.md"))
total = 0
touched = 0
for f in files:
    md = f.read_text(encoding="utf-8")
    out, count = convert_markdown(md)
    if out != md:
        f.write_text(out, encoding="utf-8")
        touched += 1
        total += count
        print(f"{count}\t{f.as_posix()}")
print(json.dumps({"files": len(files), "touched": touched, "convertedFences": total}, ensure_ascii=False, indent=2))
